### ISSUE TITLE
Better DI support 

### ISSUE THREAD

**Issue by JeremySkinner** [JeremySkinner/FluentValidation#806]
At the moment DI support is limited to creation of root validators. The container is abstracted behind IValidatorFactory

We should allow resolution of child validators and other components during
Validation. To do this a resolver should be exposed on ValidationContext that calls into the container. New overloads of SetValidator should make use of this.

**Comment by waynebrantley** [comment]
Naively support .net core DI to handle this.
Optionally, create a package that integrates the native DI into your stuff with a configure call, etc.

**Comment by JeremySkinner** [comment]
@waynebrantley  Supporting native .net core DI directly isn't a good idea as it's not supported on all platforms that FluentValidation supports. 

The plan is to use the existing `IValidatorFactory` to do this, which can then wrap a container of your choosing. For `FluentValidation.AspNetCore` this will be provided by the asp.net core service provider. For other platforms will provide a simplistic implementation by default which can then be swapped out.

**Comment by cwe1ss** [comment]
I'd also love to see a separate package for `Microsoft.Extensions.DependencyInjection` that allows us to use it outside of ASP.NET Core.

Great project! 👍

**Comment by nsarris** [comment]
I've written a set of IoC Extensions here: 

https://github.com/nsarris/FluentValidation.IoC

It's an abstract solution and I have implemented an IoC provider with Unity. It's very easy to add one for Microsoft.Extensions.DependencyInjection. The solution supports net 46/47/standard/core

Please check the tests for usage. Unfortunately I didn't have time for more tests or proper documentation

FluentValidation's architecture has Validators building their logic upon construction, so I tried to build the DI extensions around the idea of them being singletons, to avoid the build up overhead.

The validation state is held within the Context classes so the whole idea is to attach an IValidatorFactory and a custom IDependencyResolver in the context. For this I used the Dictionary for custom data. It would be really easy to fuse this with the FluentValidation if the context class would add those two properties.

The other thing I tried to tackle was Literal resolution. The current version of FluentValidation treats string literals in a static way, either at build up or with the LanguageManager which is static. So I introduced an ILiteralService which can be injected as well. As there is no context available in the WithMessage delegates, I had to use a ServiceLocator for the LiteralService. 

For IValidatorProvider/IDependencyResolver there is the option of injection by constructor, explicit construction or serviceLocation (if not provider)

I decided to keep IValidatorProvider and IDependncyResolver separate, although my Unity implementation provides them in the same container instance. 

I can provide more info if you are interested.

**Comment by JeremySkinner** [comment]
@nsarris thanks for the suggestion. This seems far more complex than what I'm aiming for with the DI support here, at least for the initial version. 

The plan for the initial version of the DI support is to provide a way to resolve child validators only specifying the type, eg `SetValidator<MyChildValidator>()` rather than having to instantiate the child validator, or pass it via the constructor. 

This would be done by passing an `IValidatorFactory` into the validator via the `ValidationContext`. This would be passed in via the integration with ASP.NET Core's validation provider. In time, this may be expanded to other platforms, but for now the focus on this is only for ASP.NET Core via the service provider mechanism. We already have an implemention of IValidatorFactory that works with the service provider which can be used for this. For integration with other DI containers, it'll be up to the user to provide the appropriat adapter. I'm not at all keen on providing integration with the various different containers out there, I've tried this in the past and it's a maintenance nightmare.

For future versions, we can look at providing hooks for doing this outside of web projects, but that's not the focus for now. 

That being said, I'd be happy to review your solution to see if any of this could be used for what we're aiming for. Unfortunately I don't have much free time at the moment either, and as this is a fairly sizable piece this will take me some time to review. I know you said you don't have time for documentation, but please could you at least provide a high-level overview of the architecture and usage. That'll give me a good starting point to be able to do a proper review. 

> so I tried to build the DI extensions around the idea of them being singletons, to avoid the build up overhead.

Since v6.4 the validator cache means that validators do not need to be singletons, and the asp.net cor integration intentially registers them as transient to avoid the confusion and complexities that come from singleton-scoped dependencies, which can cause a lot of gotchas especially for new devs who aren't familiar with the implications. It's important that any solution can work with transient as the default scope. 

> For this I used the Dictionary for custom data. It would be really easy to fuse this with the FluentValidation if the context class would add those two properties.

I am not keen on adding additional properties to the context. As this is going to be a separate package, using the `RootContextData` dictionary to extend the context, and then wrapping this with extension method is preferable.

 > The other thing I tried to tackle was Literal resolution

Could you provide some details of what your use case is for this? This sounds like it'd be out of scope for what we're aiming for here, at least for the initial version, or am I missing something?

**Comment by nsarris** [comment]
> That being said, I'd be happy to review your solution to see if any of this could be used for what we're aiming for. Unfortunately I don't have much free time at the moment either, and as this is a fairly sizable piece this will take me some time to review. I know you said you don't have time for documentation, but please could you at least provide a high-level overview of the architecture and usage. That'll give me a good starting point to be able to do a proper review.

Sure, I will try to put something together this weekend for you to get a better idea.

> For integration with other DI containers, it'll be up to the user to provide the appropriat adapter. I'm not at all keen on providing integration with the various different containers out there, I've tried this in the past and it's a maintenance nightmare.

I built everything around the existing IValidatorFactory and two interfaces (IDependencyResolver and ILiteralService). The unity container project is just a proof of concept implementation. You can use any container you want by inheriting those simple interfaces so I think we're on the same page here.

There's three interfaces because I tried to tackle three DI related issues.

1) The resolution of Validators by validated type or validator type at run time, through IValidatorFactory, which is what you are looking for.
2) The resolution of dependencies within the validator's payload. This requires a proper IoC Container/Resolver and a disposable wrapper around an IoC context. More on this on a later post .
3) The resolution of string literals at run time using proper DI which I prtially cover later on in this post.

> Since v6.4 the validator cache means that validators do not need to be singletons, and the asp.net cor integration intentially registers them as transient to avoid the confusion and complexities that come from singleton-scoped dependencies, which can cause a lot of gotchas especially for new devs who aren't familiar with the implications. It's important that any solution can work with transient as the default scope.

The buildup in constructor kind of contradicts this. In order for the validators to really be transient you could refactor things in the future to use rule builder classes to seperatre the rule building (singleton) from the actual validation execution (transient) . This would really help with DI as well as you could add dependencies in the executor (via constructor or otherwise). 

In the current state of the framework you can emulate this partially by using a validator for the executor (transient) with an included validator as the rule builder (singleton). Alas, this doesn't help with DI because the rule builder needs the dependencies.

> I am not keen on adding additional properties to the context. As this is going to be a separate package, using the RootContextData dictionary to extend the context, and then wrapping this with extension method is preferable.

This is exactly how I approached it and I have no problem keeping it this way.

> Could you provide some details of what your use case is for this? This sounds like it'd be out of scope for what we're aiming for here, at least for the initial version, or am I missing something?

In FluentValidation you provide string literals in three ways:
WithName, WithMessage and WithLocalizedMessage, ValidatorOptions.PropertyNameResolver,  ValidatorOptions.DisplayNameResolver (and the ValidatorOptions.Language translations)

None of these ways are really DI friendly in theory. The WithName and WithMessage don't inject the ValidationContext and thus it's not possible to resolve anything within the delegate. The WithLocalizedMessage is based on static properties on a resource type.  Finally the PropertyNameResolver/DisplayNameResolver are statics which is ServiceLocation not proper DI (plus there is no ErrorMessageResolver). I implemented ILiteralService to move towards a proper DI solution for string literals. I did have to implement it with Service Location for now but I was hoping to talk you into adding an overload to WithName/WithMessage that injects the context into the delegate.

The service looks like this:

````
public interface ILiteralService
{
    string GetPropertyName(Type entityType, string propertyName);
    string GetPropertyName<T>(string propertyName);
    string GetValidationErrorMessage(string code);
}
````

This is used like this:

````
RuleFor(x => x.Id)
.NotEmpty().ResolveName()
.WithErrorCode("VatValidationServiceFailure")
.ResolveMessage();
````

ResolveName and ResolveMessage are two of the extension methods in the IoC package. The encapsulate the LiteralService (currently retrieved by a static ServiceLocator) to provide literal resolution at runtime (for property names and error messages respectively).

On a sidenote, a minor issue I faced here was that I had to use reflection to the get the error code from the ErrorCodeSource built from WithErrorCode. (RuleBuilder.Rule.CurrentValidator.Options.ErrorCodeSource._message)

**Comment by JeremySkinner** [comment]
> Sure, I will try to put something together this weekend for you to get a better idea.

Thanks. 

> The buildup in constructor kind of contradicts this. In order for the validators to really be transient you could refactor things in the future to use rule builder classes to seperatre the rule building (singleton) from the actual validation execution (transient)

Agreed, I’ve looked at doing this before but this is such a massive breaking change that really has very little benefit for the vast majority of users, and introduces several other complexities regarding how the rules interact with each other. I spent quite a while prototyping the feasibility of this and in the end decided to keep with the current approach.

Some thoughts regarding ILiteralService:

- I think this would be better split into different resolver classes for each job rather than a single service. This keeps each class focused on a single task. 
- no need for both generic and non generic versions of GetPropertyName. The usual approach with FluentValidation is the public api uses generics and internal api (which this would be) doesn’t. 
- GetPropertyName should probably be GetDisplayName, and there should be a separate GetPropertyName (to correspond to WithName and OverridePropertyName respectively)
- I think we can get rid of the ResolveName method. I’d imagine that this would be the default behaviour, falling back to the default message if a custom resolver is not specified. 
- I don’t really like GetValidationErrorMessage being tied to the error code. Error codes and messages are intentionally separate, and I don’t like the idea of having the set an error code to be able to find the default message for the validator. Again I think this should be a separate resolver class that allows you to provide the message for a specific validator, or it would fall back to use the LanguageManager. If you do opt to provide a custom resolver, then I could imagine that you could use this to implement the error code-based behaviour if you wanted to, but I wouldn’t want to provide this as a default api method. 
- if these services are going to be used internally by FV core then there will also be a need to specify them if the user isn’t using a container. Probably via a global option, maybe replacing the current global DisplayNameResolver/PropertyNameResolver methods with instances that can be swapped out. Not ideal but we do need to support that scenario. Another idea maybe to provide a default, transparent ioc implementation for FV’s own services (maybe by embedding TinyIoc, or something), which automatically gets swapped out with the ServiceProvider when running under AspNetCore. 

Thanks again for this, you’ve given me a lot to think about.

**Comment by JeremySkinner** [comment]
Hi, I'm afraid there are no current plans to support open generics in this way, although if you or someone else wanted to submit a pull request to add it I'd be happy to give it a review.

**Comment by nsarris** [comment]
> There's three interfaces because I tried to tackle three DI related issues.
> 1. The resolution of Validators by validated type or validator type at run time, through IValidatorFactory, which is what you are looking for.
> 2. The resolution of dependencies within the validator's payload. This requires a proper IoC Container/Resolver and a disposable wrapper around an IoC context. More on this on a later post .
> 3. The resolution of string literals at run time using proper DI which I prtially cover later on in this post.

Just to wrap No.2 up as I promised. I think we're on the same page with No.1 and No.3 revolves around injecting the context in the validation rule delegate.

No.2 is about addressing the need to resolve dependencies within the validation delegates. Suppose that to validate a VAT number you have a dependency on a service with side effects that calls a goverment API to fetch data. How would you inject that dependency into the validator. One option is to have the validator created by a DI container and inject the dependency in the validators constructor. The problem with this approach is that the validator needs to be instantiated each time and that will cost you the rule buildup. We discussed this above and you mentioned this is by design. However I tried to provide an option for the validators to be singletons (so no buildup cost) and the dependency resolver to be injected in the context, which in turn injects it into the validator delegate.

To implement this i used a simple IDependencyResolver interface that you can use to create a proxy to any container and inject it into the context. For the implementation of the context buildup I used an IoCValidationContext class that uses a IValidatorFactory and IDependencyResolver as dependencies and goes on to create the ValidationContext and provide an interface for Validation execution. 

The IoCValidationContext also handles lifecylcle by IDisposable, by simple propagating Dispose to the IDependencyResolver implementation.

This is the basic idea around it. If you want me to further document my project I think I can find some time to do it soon.

**Comment by nsarris** [comment]
FYI, I took another look and I found out that proper dependency injection for name and message resolution can be achieved using the Configure method and DisplayName/PropertyName/MessageBuilderContext (which gets rid of the ServiceLocation approach)

For the first two I created a custom ILiteralSource class from which you can extract the ILiteralService dependency as the GetString function is called with a ValidationContext.

For the later, the MessageBuilderContext contains a reference to the parent ValidationContext, from where you can get validation time dependencies.

**Comment by JeremySkinner** [comment]
Yes, that definitely helps to simplify it a bit. 

The only thing to keep in mind is that a `MessageBuilder` is associated with the rule, not with the individual validator, so this will not do what you'd expect it to:

```
// Only the `bar` message will ever be used. 
RuleFor(x => x.Foo).NotNull().ResolveMessage("foo").Must(Whatever).ResolveMessage("bar");
``` 

It'd be better to set the `ErrorMessageSource`, as this is per-validator. You can again use a StringSource for this:

```csharp
return rule.Configure(config => {
	config.CurrentValidator.Options.ErrorMessageSource = new MySource(code, ...); 
});
```

Somewhat related to this, I recently committed a change that allows you to retrive the message using an error code in a very smiliar way to this, but the message is retrieved from the `LanguageManager`. 

For injecting validators, rather than using a new method that takes a `Type`, I was thinking instead of a static method that allows you to specify the validator type:

```csharp
RuleFor(x => x.Address).SetValidator(Resolve.Validator<Address, AddressValidator>());
```

`Resolve.Validator` would just return a delegate, meaning it can be passed to the existing `WithMessage(Func<T,TValidator>)`, and is strongly typed. What do you think of that syntax?

**Comment by nsarris** [comment]
> It'd be better to set the ErrorMessageSource, as this is per-validator. You can again use a StringSource for this:

Worked like a charm, thanks this is much cleaner.

> Somewhat related to this, I recently committed a change that allows you to retrive the message using an error code in a very smiliar way to this, but the message is retrieved from the LanguageManager.

Yes, I noticed but that still requires setting up a static resource class. This is acceptable in certain cases but it in general it violates most of the principles of DI/IoC

> For injecting validators, rather than using a new method that takes a Type, I was thinking instead of a static method that allows you to specify the validator type:
Resolve.Validator would just return a delegate, meaning it can be passed to the existing WithMessage(Func<T,TValidator>), and is strongly typed. What do you think of that syntax?

Like I said above statics and DI are contradicting patterns. In my humble opinion statics should be considered an anti-pattern nowdays, except for pure functions in the sense of helpers and extensions methods. Statics create problems with unit testing, break inversion of control and do not allow any other lifecycle except singleton and do not support DI scoping. I sencerily advise against it. I wanted to suggest the same for the validator options but I was happy to see you already decided to change it.

In the suggestion above the static "Resolve" class would essentially be a staic ServiceLocator for the IoC container/ Dependency resolver. I believe that instead of setting up the resolver statically you should inject it in the ValidationContext (which is what I did). Your ValidationContext carries the state of the validation process. If you inject your resolver in the context(or a leaner wrapper around it using an interface like I did) you have access to it in all steps of the process, including Validator resolution.

So then you can resolve either by spefic validator (AddressValidator) or generic type (Address) or with any kind of inteface mapping provider by the framework choice.

**Comment by JeremySkinner** [comment]
> In the suggestion above the static "Resolve" class would essentially be a staic ServiceLocator for the IoC container/ Dependency resolver.

No, it wouldn’t....sorry I don’t think I explained it clearly. The method would not act as a service locator - it would simply return a *delegate* that would use the validatoncontext for injection - no service location required. I’ll try and throw up an example at some point.

**Comment by nsarris** [comment]
Ok I see what you mean know and it sounds promising

**Comment by JeremySkinner** [comment]
I've settled on the following syntax. `InjectValidator()` will infer the child validator type to inject:

```csharp
RuleFor(x => x.Address).InjectValidator();
```

Assuming the Address property is of type `Address`, InjectValidator will resolve an `IValidator<Address>` from the service provider.

For more explicit control over the type to resolve, the method can optionally take a lambda which receives the service provider and the context:

```csharp
RuleFor(x => x.Address).InjectValidator((s, ctx) => s.Resolve<MyCustomAddressValidator>());
```

The service provider will be of type `IServiceProvider` (no need for a custom `IDependencyResolver` interface here), as this is part of the framework already, and is implemented by the .NET Core built-in container (meaning no additional dependency or package required for the common interface). 

For 8.2 these new methods will be part of the `FluentValidation.AspNetCore` package, and can potentially be split out into a separate package in the future. 

Resolving string literals for name/message as well as other services (like LanguageManager) won't be in this release as this still needs further planning, but will look at this for the future.

I won't be including an alternative of the IoCValidationContext for scoping/disposal. I strongly feel that scoping and disposal should be a feature the container should handle and FluentValidation itself should be completely agnostic of this. In a web scenario, the per-http-request lifetime would be handled transparently and this will all work correctly provided validators and their dependencies have their scopes set correctly when registered.

**Comment by waynebrantley** [comment]
@JeremySkinner if ``Address`` implements ``IAddressProvider``, potentially that is what I want resolved.
Perhaps an overload of ``RuleFor(x => x.Address).InjectValidator<IAddressProvider>();``

**Comment by JeremySkinner** [comment]
@waynebrantley You can't specify type parameters like that, as the method will already have type parameters for `T` and `TProperty`. In this case you'd use the second overload.
```
RuleFor(x => x.Address).InjectValidator((s, ctx) => s.GetService<IValidator<IAddressProvider>>());
```

**Comment by nsarris** [comment]
> The service provider will be of type IServiceProvider (no need for a custom IDependencyResolver interface here), as this is part of the framework already, and is implemented by the .NET Core built-in container (meaning no additional dependency or package required for the common interface).

Agreed, I was going to do this myself and add the GetService<T> with an extension method for .net only. 

> For 8.2 these new methods will be part of the FluentValidation.AspNetCore package, and can potentially be split out into a separate package in the future.

Do you mean you'll have seperate .net and core implementations or remove it for .net alltogether?

> Resolving string literals for name/message as well as other services (like LanguageManager) won't be in this release as this still needs further planning, but will look at this for the future.

I've covered our needs with the IoC extensions project. I'd gladly help if needed. I would only like to point out that InjectMessage(Func<IServiceProvider, IValidationContext, string>) and InjectPropertyName(Func<IServiceProvider, IValidationContext, string>) similar to InjectValidator would cover most cases I can think of

> I won't be including an alternative of the IoCValidationContext for scoping/disposal. I strongly feel that scoping and disposal should be a feature the container should handle and FluentValidation itself should be completely agnostic of this. In a web scenario, the per-http-request lifetime would be handled transparently and this will all work correctly provided validators and their dependencies have their scopes set correctly when registered.

I used IoCValidationContext for three reasons:

1. To remove all transient dependencies from the validators and move them to the IoCValidationContext instead, thus the ServiceProviders (DependencyResolver and IValidatorFactory)  are present in the IoCValidationContext ctor and not the validator ctor. (I understand you designed the validators as transient but I tried an approach that allows them to be used as singletons too.)

2. To abstract the building of an actual ValidationContext and the injection of the ServiceProviders into its RootContext. I guess this is simplified by the DependencyResolver removed (and IValidatorFactory replaced by IServiceProvider) but I'm missing how the IServiceProvider is acually injected into the flow.

3. To allow for the encapsulated ServiceProviders to be disposed. This was only to cover the case when the injected service providers scope was for the validation only. I agree this is an edge case and it could be handled manyally outside the scope of the validation process.

**Comment by JeremySkinner** [comment]
> Do you mean you'll have seperate .net and core implementations or remove it for .net alltogether?

Not sure I understand the question. This will be part of the `FluentValidation.AspNetCore` package, which is a netstandard2.0 library. ASP.NET Core runs on both .NET Core and the older .NET Framework. This initial implementation is for web projects only. 

> I would only like to point out that InjectMessage(Func<IServiceProvider, IValidationContext, string>) and InjectPropertyName(Func<IServiceProvider, IValidationContext, string>) similar to InjectValidator would cover most cases I can think of

I feel that injecting string literals is not going to be a common scenario, so I'd like to be able to provide hooks for this so you (or others) can do this, without it needing to be part of the public API. I feel a custom `IStringSource` is the best way to do this, which you can make use of if you want to.

 > I'm missing how the IServiceProvider is acually injected into the flow.

For this initial (web-only) implementation, it'll automatically be injected into the context by the `ModelValidatorProvider` which is invoked transparently by MVC's validation infrastructure. 

Initial code is here: https://github.com/JeremySkinner/FluentValidation/commit/fcff176f171f1d8bbd85e716a5bdf3699ca44bdb

**Comment by nsarris** [comment]
I have rivisted and completely refactored https://github.com/nsarris/FluentValidation.IoC based on your 8.2 implementation. At it's current form it's basicly an in place replacement of your FluentValidation.DepedencyInjection project (minus the ASP.NET stuff)

Per your suggestion I've moved to Microsoft.Extensions.DependencyInjetion.Abstractions for DI and removed the DependencyResolver and the notions of scoping/lifecycle. and kept all your semantics in (e.g. InjectValidator). So it's basicly an augmented version of your implementation. I can send a PR that merges into FluentValidation.DepedencyInjection if you want.

If you want the check the source first here are the key points:

1. Syntax
There are a lot of extensions to make life easier, a lot of "Custom" overloads to reduce boilerplate and a UseDependecies() builder extension to make injecting syntax cleaner (check the Test.Common validators)

2. Validator factory. I had to create an augmented IValidatorFactory to add support for specific validator resolution (e.g. not IValidator<T> but TValidator). I know it's a breaking change but I'm not aware of how many people acutaly use this interface, but it would be usefull to augment it. I also provide a default implementation which is a wrapper around the service provider (to facilitate injection automations).

3. DI tools: Added extensions for IServiceCollection and reflection based auto discovery of validators and scrutor based auto registration of validators.

4. ValidationContextProvider (previously IoCValidationContext). The previous version was disposable to handle scoping and lifecycle and thus a context. Since this was stripped out, this is now more clean as it's only a ValidationContextProvider/Builder/Factory. This also wraps around the service provider and is DI friendly (only needs an IServiceProvider).  It also comes with a lot of syntactic goodies to make life easier.

5. Literals. This is now a seperate set of extensions and can be removed, moved to a different namespace or be a seperate addon on its own.

Let me know what you think!

**Comment by jzabroski** [comment]
@nsarris I can't tell what your tests are testing / trying to test.  I suggest you break each assert into its own test case, and improve the naming of your tests so its not simply called `Test`.

```csharp
    [TestFixture]
    class Tests
    {
        [Test]
        public void Test()
        {
```

**Comment by nsarris** [comment]
@jzabroski 
This is mostly proof of concept, it doesn't have proper testing and no documentation at all. I will try to put something together if it makes that much difference. I merely posted this early version here to see what Jeremy thinks so I can decide how to go through with it. Even so these very basic tests (I put them together in like 10 mins) have around 60% code coverage (it's all extensions, there is no actual execution logic, just wiring)

**Comment by JeremySkinner** [comment]
Hi @nsarris thanks for the update. It'll take me a while to get through all of this, but here's some initial thoughts. 

I think we can merge some of this into the core with changes, but a lot of it I feel is over and above what I'd be prepared to support so would be better off as either something you continue to use in your own project, or a separate library if you feel like providing support for that. 

> DI tools: Added extensions for IServiceCollection and reflection based auto discovery of validators and scrutor based auto registration of validators.

I'd like to get extension methods on IServiceCollection for registering validators into the core, but it'll need some changes. At the moment we provide methods for registering validators as part of the call to `AddFluentValidation`:

```csharp
services.AddMvc()..AddFluentValidation(fv => {
   fv.RegisterValidatorsFromAssemblyContaining<MyValidator>();
   // or:
   fv.RegisterValidatorsFromAssemblyContaining(typeof(MyValidator));
   // or:
   fv.RegisterValidatorsFromAssembly(typeof(MyValidator).Assembly);
});
```

Let's make the `IServiceCollection` extension methods follow the same convention.  (3 methods, `RegisterValidatorsFromAssemblyContaining<T>`, `RegisterValidatorsFromAssemblyContaining(Type)`, `RegisterValidatorsFromAssembly(assembly)`.). 

Internally, these would use the existing `FluentValidation.AssemblyScanner` to do the work of locating the validators instead of `ReflectionHelper.AutoDiscoverValidatorTypes`. 

I'm not keen on merging in the behaviour of `ReflectionHelper.AutoDiscoverValidatorTypes` - I feel that this level of mapping is beyond what FluentValidation should support out the box, and is very much heading into the territory of DI containers that perform automapping. This really is out of scope of what I want to support - at the end of the day, FluentValidation is a validation library not a DI library and I don't want to start having to support a whole load of code that isn't related to the core domain of validation. I'm also not keen on bringing across the `ServiceLocator` class for the same reasons. However I would be open to adding in extensibility points to `AssemblyScanner` to allow the logic to be pluggable, so you can swap out the default search behaviour if you want to.  

Also, the default service lifetime for validators should always be `Transient`. Although singleton is better for performance, it adds a huge amount of confusion and complexity with ensuring that singleton validators don't depend on non-singleton (or web-scoped) services. This is a common pitfall for many developers. Making the default lifetime transient avoids this (and those that really know what they're doing can of course use singleton instead).

> Validator factory. I had to create an augmented IValidatorFactory to add support for specific validator resolution

I don't think we'll bring this into the core as 1) it's a breaking change and 2) it'd only be necessary if we're for more complex auto registrations. If you need this behaviour, I'd suggest resolving the validators from the container directly and bypass the factory. 

> Literals. This is now a seperate set of extensions and can be removed, moved to a different namespace or be a seperate addon on its own.

Looks good, although as I've mentioned in previous posts I don't think the literal injection is something we want to support in the core project as this isn't a common scenario, and we provide the necessary extension points to do this.

So if you're interested in putting together a PR for the service collection extensions, then the first step would be the 3 Register* methods that delegate to AssemblyScanner to do the search and register the results with the container as transient by default.

A second step as a separate PR would be to add extension points to allow the AssemblyScanner's logic to be pluggable.

**Comment by nsarris** [comment]
Thanks for the swift reply

> I'm not keen on merging in the behaviour of `ReflectionHelper.AutoDiscoverValidatorTypes` - I feel that this level of mapping is beyond what FluentValidation should support out the box, and is very much heading into the territory of DI containers that perform automapping. This really is out of scope of what I want to support - at the end of the day, FluentValidation is a validation library not a DI library and I don't want to start having to support a whole load of code that isn't related to the core domain of validation. I'm also not keen on bringing across the `ServiceLocator` class for the same reasons. However I would be open to adding in extensibility points to `AssemblyScanner` to allow the logic to be pluggable, so you can swap out the default search behaviour if you want to.

I completely missed the existence of AssemblyScanner in the core. In fact I was really reluctant to implement any reflection code in my code as well for the same reasons you describe. I'll revert to AssemblyScanner.

The ServiceLocator I forgot to mention was supposed to be plugged into the static options if you'd want to merge it, to support non DI driven environments. I'm not really keen on service location either so I'm ok with ridding it.

> Also, the default service lifetime for validators should always be `Transient`. Although singleton is better for performance, it adds a huge amount of confusion and complexity with ensuring that singleton validators don't depend on non-singleton (or web-scoped) services. This is a common pitfall for many developers. Making the default lifetime transient avoids this (and those that really know what they're doing can of course use singleton instead).

You're right about the default lifetime, I was thinking that with everything being injected through the context the validators wouldn't have any dependencies, but yes it's a pitfall. I'll revert to Transient.

> I don't think we'll bring this into the core as 1) it's a breaking change and 2) it'd only be necessary if we're for more complex auto registrations. If you need this behaviour, I'd suggest resolving the validators from the container directly and bypass the factory.

Here's what got me buffled during my design. Is there a point to IValidatorFactory when you have DI? If yes then we should only register the IValidatorFactory. If not not then we should register validators so what's the point of IValidatorFactory? My current implementation registers validators and simply implements IValidatorFactory as a wrapper around IServiceProvider. I chose to get resolution through the factory partly because that's what I saw in your 8.2 code and partly because I'd hoped you'd agree to augment IValidatorFactory. So my suggestion is this: Use the service provider under all the InjectValidator methods and leave IValidatorFactory for the SetValidator methods. It's clear semanticaly and provides a clean layer on top of the ServiceProvider.

> Looks good, although as I've mentioned in previous posts I don't think the literal injection is something we want to support in the core project as this isn't a common scenario, and we provide the necessary extension points to do this.

Like I said this can work as a separate library. If you want to endorse it, add a seperate repo in the FluentValidator family and I could PR it there.

> So if you're interested in putting together a PR for the service collection extensions, then the first step would be the 3 Register* methods that delegate to AssemblyScanner to do the search and register the results with the container as transient by default. A second step as a separate PR would be to add extension points to allow the AssemblyScanner's logic to be pluggable.

Sure I can do that. For the AssemblyScanner to be pluggable there are three options. Inject it (fall back to default), service locate it (through static options) and provide it in the IServiceCollection extension (e.g. services.AddValidators(assemblyScanner)). All of the above can coexist. What's your pick?

All of the above cover evertything except all the rulebuilder extension methods I built for ease of use. What's your thoughts on them? Do you want some examples/documentation? Do you prefer to place them in seperate repo or maybe namespace?

**Comment by JeremySkinner** [comment]
> Is there a point to IValidatorFactory when you have DI?

Not as much these days - it predates the built in ASP.NET ServiceCollection. But it still has its uses. For example, some users still rely on the `[Validator(typeof(MyValidator))]` attributes to wire up validators to models, rather than using the container, which is legacy but still supported (via the AttributedValidatorFactory). It allows instantiation of validators to be external to the regular DI process. Personally I wouldn't recommend using anything other than the container to instantiate validators, but the old approach is still supported (at least until FluentValidation 9 anyway). The default implementation of `IValidatorFactory` now just requests the container to resolve the service. 

Starting with FluentValidation 9, I'll *probably* drop `IValidatorFactory` from the validation pipeline for ASP.NET Core, but the legacy ASP.NET support (MVC5/WebApi2) will continue to use `IValidatorFactory` as there's no native DI there (I may also just drop MVC5/WebApi2 support completely, but not decided yet) 

> So my suggestion is this: Use the service provider under all the InjectValidator methods and leave IValidatorFactory for the SetValidator methods. It's clear semanticaly and provides a clean layer on top of the ServiceProvider.

I'll have a think about that and get back to you, but if IValidatorFactory is going away with 9.x anyway, it may make sense to just leave it as-is for 8.x

> All of the above can coexist. What's your pick?

I think I'd like to keep it as simple as possible. As the scanner is only used during startup, I want to avoid having to deal with injecting anything here. My initial thought would be to just allow passing a delegate to the AssemblyScanner. When specifed, this delegate would be executed to perform the scan rather than using the built-in logic. The service collection extension methods, would then allow this callback as an optional argument, passing it to the scanner. 

> All of the above cover evertything except all the rulebuilder extension methods I built for ease of use. What's your thoughts on them

Haven't had a chance to look at that yet, will try early next week.

**Comment by JeremySkinner** [comment]
I've released 8.2 with the dependency injection extensions & service collection extensions. It should show up in nuget shortly. I'm going to close this issue - please open new issues for future enhancements to this.
